<!DOCTYPE html>
<html>
<head>
    
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>SEforSDL - C++ to Java: Exceptions</title>
    <link rel="stylesheet" href="..\..\markbind\css\bootstrap.min.css">
    <link rel="stylesheet" href="..\..\markbind\css\bootstrap-vue.min.css">
    <link rel="stylesheet" href="..\..\markbind\css\font-awesome.min.css" >
    <link rel="stylesheet" href="..\..\markbind\css\bootstrap-glyphicons.min.css" >
    <link rel="stylesheet" href="..\..\markbind\css\github.min.css">
    <link rel="stylesheet" href="..\..\markbind\css\markbind.css">
    
    
    <link rel="icon" href="/se-book/favicon.ico">
</head>
<body>
<div id="app">
    <div id="content-wrapper">
  <div>
    <navbar placement="top" type="inverse">
      <a slot="brand" href="/" title="Home" class="navbar-brand">Software Engineering for Self-Directed Learners</a>
      <li slot="right">
        <searchbar :data="searchData" placeholder="Search" :on-hit="searchCallback" menu-align-right=""></searchbar>
      </li>
    </navbar>
  </div>
  <div>
    <link rel="stylesheet" href="/se-book/css/textbook.css">
    <div class="website-content" id="all">
      <p><span class="dimmed"><strong><span>C++ to Java →
</span></strong>
        </span>
      </p>
      <div id="title">
        <div>
          <h2 id="exceptions">Exceptions</h2>
        </div>
      </div>
      <div id="main">
        <div>
          <div>
            <panel type="seamless" expanded="">
              <span slot="header" class="card-title"><div>
<h4 id="what-are-exceptions">What are Exceptions?</h4></div></span>
              <div class="indented">
                <p><span class="badge badge-light"><span><span class="dimmed"><span><span class="fas fa-trophy"></span> Can explain Java Exceptions</span>
                  </span>
                  </span>
                  </span>
                  <p></p>
                </p>
                <p></p>
                <div>
                  <p>Given below is an extract from the <small><span style="opacity: 0.5"><span class="badge badge-secondary">-- Java Tutorial</span></span></small>, with slight adaptations.</p>
                  <blockquote>
                    <tip-box type="definition">
                      <div>
                        <p><strong>Exception</strong>:</p>
                        <blockquote>
                          <p>The term <em>exception</em> is shorthand for the phrase &quot;exceptional event.&quot; An <em>exception</em> is an event, which occurs during the execution of a program, that disrupts the normal flow of the program's instructions.
                            –- Java Tutorial (Oracle Inc.)</p>
                        </blockquote>
                      </div>
                    </tip-box>
                    <p><strong>When an error occurs within a method, the method creates an <em>exception object</em> and hands it off to the runtime system.</strong> The exception object contains information about the error, including its type and the state
                      of the program when the error occurred. Creating an exception object and handing it to the runtime system is called <em>throwing</em> an exception.</p>
                    <p><strong>After a method throws an exception, the runtime system attempts to find something to handle it in the <tooltip content="the ordered list of methods that had been called to get to the method where the error occurred"><em>call stack</em></tooltip>.</strong>                      The runtime system searches the call stack for a method that contains a block of code that can handle the exception. This block of code is called an <em>exception handler</em>. The search begins with the method in which the error
                      occurred and proceeds through the call stack in the reverse order in which the methods were called. When an appropriate handler is found, the runtime system passes the exception to the handler. An exception handler is considered
                      appropriate if the type of the exception object thrown matches the type that can be handled by the handler.</p>
                    <p><strong>The exception handler chosen is said to <em>catch</em> the exception.</strong> If the runtime system exhaustively searches all the methods on the call stack without finding an appropriate exception handler, the program terminates.</p>
                    <p><strong>Java code must honor the <em>Catch or Specify</em> Requirement.</strong> This means that code that might throw certain exceptions must be enclosed by either of the following:</p>
                    <ul>
                      <li>A <code>try</code> statement that catches the exception. The <code>try</code> must provide a handler for the exception.</li>
                      <li>A method that specifies that it can throw the exception. The method must provide a <code>throws</code> clause that lists the exception.</li>
                    </ul>
                    <p><strong>There are three basic categories of exceptions, only one of which is subject to the 'Catch or Specify' Requirement.</strong></p>
                    <ul>
                      <li><strong><em>Checked</em> exceptions</strong>: These are exceptional conditions that a well-written application should anticipate and recover from. For example, suppose an application prompts a user for an input file name, then opens
                        the file by passing the name to the constructor for java.io.FileReader. Normally, the user provides the name of an existing, readable file, so the construction of the <code>FileReader</code> object succeeds, and the execution of
                        the application proceeds normally. But sometimes the user supplies the name of a nonexistent file, and the constructor throws <code>java.io.FileNotFoundException</code>. A well-written program will catch this exception and notify
                        the user of the mistake, possibly prompting for a corrected file name.
                        <ul>
                          <li><mark>Checked exceptions are subject to the Catch or Specify Requirement.</mark></li>
                          <li>All exceptions are checked exceptions, except for <code>Error</code>, <code>RuntimeException</code>, and their subclasses.</li>
                        </ul>
                      </li>
                    </ul>
                    <p></p>
                    <ul>
                      <li><strong>Errors</strong>: These are exceptional conditions that are external to the application, and that the application usually cannot anticipate or recover from. For example, suppose that an application successfully opens a file
                        for input, but is unable to read the file because of a hardware or system malfunction. The unsuccessful read will throw <code>java.io.IOError</code>. An application might choose to catch this exception, in order to notify the user
                        of the problem — but it also might make sense for the program to print a stack trace and exit.
                        <ul>
                          <li>Errors are not subject to the 'Catch or Specify' Requirement.</li>
                          <li>Errors are those exceptions indicated by <code>Error</code> and its subclasses.</li>
                        </ul>
                      </li>
                    </ul>
                    <p></p>
                    <ul>
                      <li><strong>Runtime exceptions</strong>: These are exceptional conditions that are internal to the application, and that the application usually cannot anticipate or recover from. These usually indicate programming bugs, such as logic
                        errors or improper use of an API. For example, consider the application described previously that passes a file name to the constructor for FileReader. If a logic error causes a null to be passed to the constructor, the constructor
                        will throw <code>NullPointerException</code>. The application can catch this exception, but it probably makes more sense to eliminate the bug that caused the exception to occur.
                        <ul>
                          <li>Runtime exceptions are not subject to the 'Catch or Specify' Requirement.</li>
                          <li>Runtime exceptions are those indicated by <code>RuntimeException</code> and its subclasses.</li>
                        </ul>
                      </li>
                    </ul>
                    <p><strong>Errors and runtime exceptions are collectively known as <em>unchecked</em> exceptions.</strong></p>
                  </blockquote>
                </div>
                <div></div>
              </div>
            </panel>
          </div>
          <div>
            <panel type="seamless" expanded="">
              <span slot="header" class="card-title"><div>
<h4 id="how-to-use-exceptions">How to Use Exceptions</h4></div></span>
              <div class="indented">
                <p><span class="badge badge-light"><span><span class="dimmed"><span><span class="fas fa-trophy"></span> Can use Java Exceptions</span>
                  </span>
                  </span>
                  </span>
                  <p></p>
                </p>
                <p></p>
                <div>
                  <p>Given below is an extract from the <small><span style="opacity: 0.5"><span class="badge badge-secondary">-- Java Tutorial</span></span></small>, with slight adaptations.</p>
                  <blockquote>
                    <p><strong>A program can catch exceptions by using a combination of the <code>try</code>, <code>catch</code>, and <code>finally</code> blocks.</strong> Here is an example:</p>
                    <pre><code class="hljs java"><span class="hljs-function"><span class="hljs-keyword">public</span> <span class="hljs-keyword">void</span> <span class="hljs-title">writeList</span><span class="hljs-params">()</span> </span>{
    PrintWriter out = <span class="hljs-keyword">null</span>;

    <span class="hljs-keyword">try</span> {
        System.out.println(<span class="hljs-string">"Entering"</span> + <span class="hljs-string">" try statement"</span>);

        out = <span class="hljs-keyword">new</span> PrintWriter(<span class="hljs-keyword">new</span> FileWriter(<span class="hljs-string">"OutFile.txt"</span>));
        <span class="hljs-keyword">for</span> (<span class="hljs-keyword">int</span> i = <span class="hljs-number">0</span>; i &lt; SIZE; i++) {
            out.println(<span class="hljs-string">"Value at: "</span> + i + <span class="hljs-string">" = "</span> + list.get(i));
        }
    } <span class="hljs-keyword">catch</span> (IndexOutOfBoundsException e) {
        System.err.println(<span class="hljs-string">"Caught IndexOutOfBoundsException: "</span> + e.getMessage());

    } <span class="hljs-keyword">catch</span> (IOException e) {
        System.err.println(<span class="hljs-string">"Caught IOException: "</span> +  e.getMessage());

    } <span class="hljs-keyword">finally</span> {
        <span class="hljs-keyword">if</span> (out != <span class="hljs-keyword">null</span>) {
            System.out.println(<span class="hljs-string">"Closing PrintWriter"</span>);
            out.close();
        } <span class="hljs-keyword">else</span> {
            System.out.println(<span class="hljs-string">"PrintWriter not open"</span>);
        }
    }
}
</code></pre>
                    <ul>
                      <li>
                        <p>The <code>try</code> block identifies a block of code in which an exception can occur.</p>
                      </li>
                      <li>
                        <p>The <code>catch</code> block identifies a block of code, known as an exception handler, that can handle a particular type of exception.</p>
                      </li>
                      <li>
                        <p>The <code>finally</code> block identifies a block of code that is guaranteed to execute, and is the right place to close files, recover resources, and otherwise clean up after the code enclosed in the try block.</p>
                      </li>
                      <li>
                        <p>The <code>try</code> statement should contain at least one <code>catch</code> block or a finally block and may have multiple <code>catch</code> blocks.</p>
                      </li>
                      <li>
                        <p>The class of the exception object indicates the type of exception thrown. The exception object can contain further information about the error, including an error message.</p>
                      </li>
                    </ul>
                    <p><strong>You can use the <code>throw</code> statement to throw an exception.</strong> The throw statement requires a single argument: a throwable object. Throwable objects are instances of any subclass of the <code>Throwable</code>                      class. Here's an example of a throw statement.</p>
                    <pre><code class="hljs java"><span class="hljs-keyword">if</span> (size == <span class="hljs-number">0</span>) {
    <span class="hljs-keyword">throw</span> <span class="hljs-keyword">new</span> EmptyStackException();
}
</code></pre>
                    <p><strong>If the a method doesn't catch the checked exceptions that can occur within it, it must specify that it can throw these exceptions</strong> using a <code>throws</code> clause. Here's an example.</p>
                    <pre><code class="hljs java"><span class="hljs-function"><span class="hljs-keyword">public</span> <span class="hljs-keyword">void</span> <span class="hljs-title">writeList</span><span class="hljs-params">()</span> <span class="hljs-keyword">throws</span> IOException, IndexOutOfBoundsException </span>{
    PrintWriter out = <span class="hljs-keyword">new</span> PrintWriter(<span class="hljs-keyword">new</span> FileWriter(<span class="hljs-string">"OutFile.txt"</span>));
    <span class="hljs-keyword">for</span> (<span class="hljs-keyword">int</span> i = <span class="hljs-number">0</span>; i &lt; SIZE; i++) {
        out.println(<span class="hljs-string">"Value at: "</span> + i + <span class="hljs-string">" = "</span> + list.get(i));
    }
    out.close();
}
</code></pre>
                  </blockquote>
                </div>
                <div></div>
              </div>
            </panel>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
<div id="flex-div"></div>
<footer>
  <div class="text-center">
    <small>This site was built with <a href="https://markbind.github.io/markbind/">MarkBind 1.12.0</a> at Sat, 11 Aug 2018 08:12:55 GMT.</small></div>
</footer>
</div>
</body>
<script src="..\..\markbind\js\vue.min.js"></script>
<script src="..\..\markbind\js\vue-strap.min.js"></script>
<script src="..\..\markbind\js\polyfill.min.js"></script>
<script src="..\..\markbind\js\bootstrap-vue.min.js"></script>
<script>
    const baseUrl = '/se-book'
</script>
<script src="..\..\markbind\js\setup.js"></script>
</html>
